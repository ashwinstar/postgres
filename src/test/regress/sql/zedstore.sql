-- simple tests to iteratively build the zedstore
-- create and drop works
create table t_zedstore(c1 int, c2 int, c3 int) USING zedstore;
drop table t_zedstore;
-- insert and select works
create table t_zedstore(c1 int, c2 int, c3 int) USING zedstore;
insert into t_zedstore select i,i+1,i+2 from generate_series(1, 10)i;
select * from t_zedstore;
-- selecting only few columns work
select c1, c3 from t_zedstore;
-- only few columns in output and where clause work
select c3 from t_zedstore where c2 > 5;

-- Test abort works
begin;
insert into t_zedstore select i,i+1,i+2 from generate_series(21, 25)i;
abort;
insert into t_zedstore select i,i+1,i+2 from generate_series(31, 35)i;
select * from t_zedstore;

--
-- Test indexing
--
create index on t_zedstore (c1);
set enable_seqscan=off;
set enable_indexscan=on;
set enable_bitmapscan=off;

-- index scan
select * from t_zedstore where c1 = 5;

-- index-only scan
select c1 from t_zedstore where c1 = 5;

-- bitmap scan
set enable_indexscan=off;
set enable_bitmapscan=on;
select c1, c2 from t_zedstore where c1 between 5 and 10;

--
-- Test DELETE and UPDATE
--
delete from t_zedstore where c2 = 5;
select * from t_zedstore;
delete from t_zedstore where c2 < 5;
select * from t_zedstore;

update t_zedstore set c2 = 100 where c1 = 8;
select * from t_zedstore;

--
-- Test VACUUM
--
vacuum t_zedstore;
select * from t_zedstore;

--
-- Test toasting
--
create table t_zedtoast(c1 int, t text) USING zedstore;
insert into t_zedtoast select i, repeat('x', 10000) from generate_series(1, 10) i;

select c1, length(t) from t_zedtoast;

--
-- Test NULL values
--
create table t_zednullvalues(c1 int, c2 int) USING zedstore;
insert into t_zednullvalues values(1, NULL), (NULL, 2);
select * from t_zednullvalues;
select c2 from t_zednullvalues;
update t_zednullvalues set c1 = 1, c2 = NULL;
select * from t_zednullvalues;

--
-- Test COPY
--
create table t_zedcopy(a serial, b int, c text not null default 'stuff', d text,e text) USING zedstore;

COPY t_zedcopy (a, b, c, d, e) from stdin;
9999	\N	\\N	\NN	\N
10000	21	31	41	51
\.

COPY t_zedcopy (b, d) from stdin;
1	test_1
\.

COPY t_zedcopy (b, d) from stdin;
2	test_2
3	test_3
4	test_4
5	test_5
\.

COPY t_zedcopy (a, b, c, d, e) from stdin;
10001	22	32	42	52
10002	23	33	43	53
10003	24	34	44	54
10004	25	35	45	55
10005	26	36	46	56
\.

select * from t_zedcopy;

--- Test for alter table add column
create table t_zaddcol(a int) using zedstore;
insert into t_zaddcol select * from generate_series(1, 3);
-- rewrite case
alter table t_zaddcol add column b int generated always as (a + 1) stored;
select * from t_zaddcol;
-- fixed length default value stored in catalog
alter table t_zaddcol add column c int default 3;
select * from t_zaddcol;
-- variable length default value stored in catalog
alter table t_zaddcol add column d text default 'abcdefgh';
select d from t_zaddcol;

create table t_zaddcol_insert (a int) using zedstore;
insert into t_zaddcol_insert values (1);
alter table t_zaddcol_insert add column b int default 3;
select * from t_zaddcol_insert;
insert into t_zaddcol_insert values (2);
--select * from t_zaddcol_insert;
--insert into t_zaddcol_insert values (3,3);
--select b from t_zaddcol_insert;
