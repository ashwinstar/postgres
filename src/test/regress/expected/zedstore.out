-- simple tests to iteratively build the zedstore
-- create and drop works
create table t_zedstore(c1 int, c2 int, c3 int) USING zedstore;
drop table t_zedstore;
-- insert and select works
create table t_zedstore(c1 int, c2 int, c3 int) USING zedstore;
insert into t_zedstore select i,i+1,i+2 from generate_series(1, 10)i;
select * from t_zedstore;
 c1 | c2 | c3 
----+----+----
  1 |  2 |  3
  2 |  3 |  4
  3 |  4 |  5
  4 |  5 |  6
  5 |  6 |  7
  6 |  7 |  8
  7 |  8 |  9
  8 |  9 | 10
  9 | 10 | 11
 10 | 11 | 12
(10 rows)

-- selecting only few columns work
select c1, c3 from t_zedstore;
 c1 | c3 
----+----
  1 |  3
  2 |  4
  3 |  5
  4 |  6
  5 |  7
  6 |  8
  7 |  9
  8 | 10
  9 | 11
 10 | 12
(10 rows)

-- only few columns in output and where clause work
select c3 from t_zedstore where c2 > 5;
 c3 
----
  7
  8
  9
 10
 11
 12
(6 rows)

-- Test abort works
begin;
insert into t_zedstore select i,i+1,i+2 from generate_series(21, 25)i;
abort;
insert into t_zedstore select i,i+1,i+2 from generate_series(31, 35)i;
select * from t_zedstore;
 c1 | c2 | c3 
----+----+----
  1 |  2 |  3
  2 |  3 |  4
  3 |  4 |  5
  4 |  5 |  6
  5 |  6 |  7
  6 |  7 |  8
  7 |  8 |  9
  8 |  9 | 10
  9 | 10 | 11
 10 | 11 | 12
 31 | 32 | 33
 32 | 33 | 34
 33 | 34 | 35
 34 | 35 | 36
 35 | 36 | 37
(15 rows)

--
-- Test indexing
--
create index on t_zedstore (c1);
set enable_seqscan=off;
set enable_indexscan=on;
set enable_bitmapscan=off;
-- index scan
select * from t_zedstore where c1 = 5;
 c1 | c2 | c3 
----+----+----
  5 |  6 |  7
(1 row)

-- index-only scan
select c1 from t_zedstore where c1 = 5;
 c1 
----
  5
(1 row)

-- bitmap scan
set enable_indexscan=off;
set enable_bitmapscan=on;
select c1, c2 from t_zedstore where c1 between 5 and 10;
 c1 | c2 
----+----
  5 |  6
  6 |  7
  7 |  8
  8 |  9
  9 | 10
 10 | 11
(6 rows)

--
-- Test DELETE and UPDATE
--
delete from t_zedstore where c2 = 5;
select * from t_zedstore;
 c1 | c2 | c3 
----+----+----
  1 |  2 |  3
  2 |  3 |  4
  3 |  4 |  5
  5 |  6 |  7
  6 |  7 |  8
  7 |  8 |  9
  8 |  9 | 10
  9 | 10 | 11
 10 | 11 | 12
 31 | 32 | 33
 32 | 33 | 34
 33 | 34 | 35
 34 | 35 | 36
 35 | 36 | 37
(14 rows)

delete from t_zedstore where c2 < 5;
select * from t_zedstore;
 c1 | c2 | c3 
----+----+----
  5 |  6 |  7
  6 |  7 |  8
  7 |  8 |  9
  8 |  9 | 10
  9 | 10 | 11
 10 | 11 | 12
 31 | 32 | 33
 32 | 33 | 34
 33 | 34 | 35
 34 | 35 | 36
 35 | 36 | 37
(11 rows)

update t_zedstore set c2 = 100 where c1 = 8;
select * from t_zedstore;
 c1 | c2  | c3 
----+-----+----
  5 |   6 |  7
  6 |   7 |  8
  7 |   8 |  9
  9 |  10 | 11
 10 |  11 | 12
 31 |  32 | 33
 32 |  33 | 34
 33 |  34 | 35
 34 |  35 | 36
 35 |  36 | 37
  8 | 100 | 10
(11 rows)

--
-- Test VACUUM
--
vacuum t_zedstore;
select * from t_zedstore;
 c1 | c2  | c3 
----+-----+----
  5 |   6 |  7
  6 |   7 |  8
  7 |   8 |  9
  9 |  10 | 11
 10 |  11 | 12
 31 |  32 | 33
 32 |  33 | 34
 33 |  34 | 35
 34 |  35 | 36
 35 |  36 | 37
  8 | 100 | 10
(11 rows)

--
-- Test toasting
--
create table t_zedtoast(c1 int, t text) USING zedstore;
insert into t_zedtoast select i, repeat('x', 10000) from generate_series(1, 10) i;
select c1, length(t) from t_zedtoast;
 c1 | length 
----+--------
  1 |  10000
  2 |  10000
  3 |  10000
  4 |  10000
  5 |  10000
  6 |  10000
  7 |  10000
  8 |  10000
  9 |  10000
 10 |  10000
(10 rows)

--
-- Test NULL values
--
create table t_zednullvalues(c1 int, c2 int) USING zedstore;
insert into t_zednullvalues values(1, NULL), (NULL, 2);
select * from t_zednullvalues;
 c1 | c2 
----+----
  1 |   
    |  2
(2 rows)

select c2 from t_zednullvalues;
 c2 
----
   
  2
(2 rows)

update t_zednullvalues set c1 = 1, c2 = NULL;
select * from t_zednullvalues;
 c1 | c2 
----+----
  1 |   
  1 |   
(2 rows)

--
-- Test COPY
--
create table t_zedcopy(a serial, b int, c text not null default 'stuff', d text,e text) USING zedstore;
COPY t_zedcopy (a, b, c, d, e) from stdin;
COPY t_zedcopy (b, d) from stdin;
COPY t_zedcopy (b, d) from stdin;
COPY t_zedcopy (a, b, c, d, e) from stdin;
select * from t_zedcopy;
   a   | b  |   c   |   d    | e  
-------+----+-------+--------+----
  9999 |    | \N    | NN     | 
 10000 | 21 | 31    | 41     | 51
     1 |  1 | stuff | test_1 | 
     2 |  2 | stuff | test_2 | 
     3 |  3 | stuff | test_3 | 
     4 |  4 | stuff | test_4 | 
     5 |  5 | stuff | test_5 | 
 10001 | 22 | 32    | 42     | 52
 10002 | 23 | 33    | 43     | 53
 10003 | 24 | 34    | 44     | 54
 10004 | 25 | 35    | 45     | 55
 10005 | 26 | 36    | 46     | 56
(12 rows)

--- Test for alter table add column
create table t_zaddcol(a int) using zedstore;
insert into t_zaddcol select * from generate_series(1, 3);
-- rewrite case
alter table t_zaddcol add column b int generated always as (a + 1) stored;
select * from t_zaddcol;
 a | b 
---+---
 1 | 2
 2 | 3
 3 | 4
(3 rows)

-- fixed length default value stored in catalog
alter table t_zaddcol add column c int default 3;
select * from t_zaddcol;
 a | b | c 
---+---+---
 1 | 2 | 3
 2 | 3 | 3
 3 | 4 | 3
(3 rows)

-- variable length default value stored in catalog
alter table t_zaddcol add column d text default 'abcdefgh';
select d from t_zaddcol;
    d     
----------
 abcdefgh
 abcdefgh
 abcdefgh
(3 rows)

